pkg/blockstorage/azure/azuredisk.go

```diff
@@ -678,39 +662,34 @@ func (s *AdStorage) SnapshotRestoreTargets(ctx context.Context, snapshot *blocks
 
 // dynamicRegionMapAzure derives a mapping from Regions to zones for Azure. Depends on subscriptionID
 func (s *AdStorage) dynamicRegionMapAzure(ctx context.Context) (map[string][]string, error) {
-	subscriptionsCLient := subscriptions.NewClientWithBaseURI(s.azCli.BaseURI)
-	subscriptionsCLient.Authorizer = s.azCli.Authorizer
-	llResp, err := subscriptionsCLient.ListLocations(ctx, s.azCli.SubscriptionID, nil)
-	if err != nil {
-		return nil, err
-	}
-	regionMap := make(map[string]map[string]struct{})
-	for _, location := range *llResp.Value {
-		regionMap[*location.Name] = make(map[string]struct{})
-	}
+	subscriptionsClient := s.azCli.SubscriptionsClient
+	regionMap := make(map[string]LocationZoneMap)
 
-	skuClient := skus.NewResourceSkusClientWithBaseURI(s.azCli.BaseURI, s.azCli.SubscriptionID)
-	skuClient.Authorizer = s.azCli.Authorizer
-	skuResults, err := skuClient.ListComplete(ctx)
-	if err != nil {
-		return nil, err
-	}
-	for skuResults.Value().Name != nil {
-		if skuResults.Value().ResourceType != nil && *skuResults.Value().ResourceType == "disks" {
-			for _, location := range *skuResults.Value().LocationInfo {
-				if val, ok := regionMap[*location.Location]; ok {
-					for _, zone := range *location.Zones {
-						val[zone] = struct{}{}
-					}
-					regionMap[*location.Location] = val
-				}
-			}
+	locationsPager := subscriptionsClient.NewListLocationsPager(s.azCli.SubscriptionID, &armsubscriptions.ClientListLocationsOptions{IncludeExtendedLocations: nil})
+	for locationsPager.More() {
+		page, err := locationsPager.NextPage(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to advance page")
 		}
-		if err = skuResults.NextWithContext(ctx); err != nil {
-			return nil, err
+		for _, location := range page.Value {
+			regionMap[*location.Name] = make(LocationZoneMap)
```

Comment:
[nil pointer] ditto on guard for `nil` pointer

Reply:
Done